File: src/lib/sec/facts.ts
import { fetchJsonCached } from './rateLimiter'
import type { XbrlFact } from '../../types'

/**
 * The companyfacts endpoint returns every us-gaap/dei XBRL fact a company has
 * ever filed. Shape (simplified):
 * {
 *   cik, entityName,
 *   facts: {
 *     "us-gaap": {
 *       "Revenues": {
 *         label, description,
 *         units: { "USD": [ { end, val, accn, fy, fp, form, filed, start? }, ... ] }
 *       }, ...
 *     },
 *     "dei": { ... }
 *   }
 * }
 */

interface RawFactValue {
  start?: string
  end: string
  val: number
  accn: string
  fy?: number
  fp?: string
  form: string
  filed: string
}

interface RawFactConcept {
  label?: string
  description?: string
  units: Record<string, RawFactValue[]>
}

interface RawCompanyFacts {
  cik: number
  entityName: string
  facts: Record<string, Record<string, RawFactConcept>>
}

export interface CompanyFacts {
  cik: string
  name: string
  raw: RawCompanyFacts
}

export async function getCompanyFacts(cik: string): Promise<CompanyFacts> {
  const url = `https://data.sec.gov/api/xbrl/companyfacts/CIK${cik}.json`
  const raw = await fetchJsonCached<RawCompanyFacts>(url)
  return { cik, name: raw.entityName, raw }
}

/**
 * Default us-gaap concepts most useful for high-level company Q&A. We try each
 * in order and pick the first one that has data, since older filings may use a
 * different tag (e.g. RevenueFromContractWithCustomerExcludingAssessedTax).
 */
export const DEFAULT_CONCEPT_GROUPS: { name: string; tags: string[]; unit?: string }[] = [
  { name: 'Revenue', tags: ['Revenues', 'RevenueFromContractWithCustomerExcludingAssessedTax', 'SalesRevenueNet'], unit: 'USD' },
  { name: 'Cost of Revenue', tags: ['CostOfRevenue', 'CostOfGoodsAndServicesSold'], unit: 'USD' },
  { name: 'Gross Profit', tags: ['GrossProfit'], unit: 'USD' },
  { name: 'Operating Income', tags: ['OperatingIncomeLoss'], unit: 'USD' },
  { name: 'Net Income', tags: ['NetIncomeLoss'], unit: 'USD' },
  { name: 'EPS (basic)', tags: ['EarningsPerShareBasic'], unit: 'USD/shares' },
  { name: 'EPS (diluted)', tags: ['EarningsPerShareDiluted'], unit: 'USD/shares' },
  { name: 'Total Assets', tags: ['Assets'], unit: 'USD' },
  { name: 'Total Liabilities', tags: ['Liabilities'], unit: 'USD' },
  { name: "Stockholders' Equity", tags: ['StockholdersEquity', 'StockholdersEquityIncludingPortionAttributableToNoncontrollingInterest'], unit: 'USD' },
  { name: 'Cash & Equivalents', tags: ['CashAndCashEquivalentsAtCarryingValue', 'Cash'], unit: 'USD' },
  { name: 'Operating Cash Flow', tags: ['NetCashProvidedByUsedInOperatingActivities'], unit: 'USD' },
  { name: 'Research & Development', tags: ['ResearchAndDevelopmentExpense'], unit: 'USD' },
]

/**
 * Pull annual facts for the given concept tags for the requested fiscal years.
 * Filters to 10-K forms when possible to avoid duplicate quarterly entries.
 */
export function extractAnnualFacts(
  facts: RawCompanyFacts,
  conceptTags: string[],
  years: number[],
): XbrlFact[] {
  const out: XbrlFact[] = []
  const us = facts.facts['us-gaap'] ?? {}
  for (const tag of conceptTags) {
    const concept = us[tag]
    if (!concept) continue
    for (const unit of Object.keys(concept.units)) {
      for (const v of concept.units[unit]) {
        if (v.form !== '10-K') continue
        if (years.length > 0 && (!v.fy || !years.includes(v.fy))) continue
        out.push({
          concept: tag,
          label: concept.label,
          unit,
          value: v.val,
          start: v.start,
          end: v.end,
          fy: v.fy,
          fp: v.fp,
          form: v.form,
          accession: v.accn,
          filed: v.filed,
        })
      }
    }
  }
  return dedupeFacts(out)
}

function dedupeFacts(facts: XbrlFact[]): XbrlFact[] {
  const seen = new Map<string, XbrlFact>()
  for (const f of facts) {
    const key = `${f.concept}|${f.fy ?? f.end}|${f.unit}`
    const prev = seen.get(key)
    if (!prev || prev.filed < f.filed) seen.set(key, f)
  }
  return [...seen.values()].sort((a, b) => (a.fy ?? 0) - (b.fy ?? 0) || a.concept.localeCompare(b.concept))
}

/**
 * Produce a compact markdown table of the default concept groups for the given
 * years, suitable to inject into an LLM prompt as numeric context.
 */
export function summarizeFactsAsMarkdown(
  ticker: string,
  facts: RawCompanyFacts,
  years: number[],
): { markdown: string; cited: XbrlFact[] } {
  const effectiveYears = years.length > 0 ? years : inferLatestYears(facts, 3)
  const rows: string[] = []
  const cited: XbrlFact[] = []

  rows.push(`**${ticker} — ${facts.entityName}** annual facts (source: SEC XBRL companyfacts)`)
  rows.push('')
  rows.push(`| Concept | Unit | ${effectiveYears.map((y) => `FY${y}`).join(' | ')} | Source |`)
  rows.push(`|---|---|${effectiveYears.map(() => '---:').join('|')}|---|`)

  for (const group of DEFAULT_CONCEPT_GROUPS) {
    const pulled = extractAnnualFacts(facts, group.tags, effectiveYears)
    if (pulled.length === 0) continue
    const byYear = new Map<number, XbrlFact>()
    for (const f of pulled) if (f.fy != null) byYear.set(f.fy, f)
    const cells = effectiveYears.map((y) => {
      const f = byYear.get(y)
      if (!f) return '—'
      cited.push(f)
      return formatValue(f.value, f.unit)
    })
    const refs = [...new Set(pulled.map((f) => f.accession))].slice(0, 1).join(', ')
    rows.push(`| ${group.name} | ${pulled[0]?.unit ?? group.unit ?? ''} | ${cells.join(' | ')} | ${refs} |`)
  }

  return { markdown: rows.join('\n'), cited }
}

function formatValue(v: number, unit: string): string {
  if (unit === 'USD' || unit === 'usd') {
    const abs = Math.abs(v)
    if (abs >= 1e9) return `$${(v / 1e9).toFixed(2)}B`
    if (abs >= 1e6) return `$${(v / 1e6).toFixed(2)}M`
    if (abs >= 1e3) return `$${(v / 1e3).toFixed(2)}K`
    return `$${v.toFixed(2)}`
  }
  if (unit.includes('shares')) return v.toLocaleString()
  return v.toString()
}

function inferLatestYears(facts: RawCompanyFacts, n: number): number[] {
  const ys = new Set<number>()
  const us = facts.facts['us-gaap'] ?? {}
  for (const tag of ['Revenues', 'RevenueFromContractWithCustomerExcludingAssessedTax', 'NetIncomeLoss']) {
    const c = us[tag]
    if (!c) continue
    for (const u of Object.keys(c.units)) {
      for (const v of c.units[u]) {
        if (v.form === '10-K' && v.fy) ys.add(v.fy)
      }
    }
  }
  return [...ys].sort((a, b) => b - a).slice(0, n).reverse()
}

File: apps/api/filing_room/sec/facts.py
from dataclasses import dataclass
from datetime import date
from typing import Any

from filing_room.models import FinancialFact
from filing_room.sec.urls import filing_url


@dataclass(frozen=True)
class ConceptDefinition:
    label: str
    candidates: tuple[str, ...]
    preferred_units: tuple[str, ...]


CONCEPTS: dict[str, ConceptDefinition] = {
    "revenue": ConceptDefinition(
        "Revenue",
        ("RevenueFromContractWithCustomerExcludingAssessedTax", "SalesRevenueNet", "Revenues"),
        ("USD",),
    ),
    "net_income": ConceptDefinition("Net income", ("NetIncomeLoss", "ProfitLoss"), ("USD",)),
    "operating_income": ConceptDefinition("Operating income", ("OperatingIncomeLoss",), ("USD",)),
    "assets": ConceptDefinition("Total assets", ("Assets",), ("USD",)),
    "liabilities": ConceptDefinition("Total liabilities", ("Liabilities",), ("USD",)),
    "cash": ConceptDefinition(
        "Cash and equivalents",
        (
            "CashAndCashEquivalentsAtCarryingValue",
            "CashCashEquivalentsRestrictedCashAndRestrictedCashEquivalents",
        ),
        ("USD",),
    ),
    "operating_cash_flow": ConceptDefinition(
        "Operating cash flow", ("NetCashProvidedByUsedInOperatingActivities",), ("USD",)
    ),
    "diluted_eps": ConceptDefinition(
        "Diluted EPS", ("EarningsPerShareDiluted",), ("USD/shares", "USD / shares")
    ),
    "shares_diluted": ConceptDefinition(
        "Diluted shares", ("WeightedAverageNumberOfDilutedSharesOutstanding",), ("shares",)
    ),
}


def _duration_days(unit: dict[str, Any]) -> int | None:
    start, end = unit.get("start"), unit.get("end")
    if not start or not end:
        return None
    return (date.fromisoformat(end) - date.fromisoformat(start)).days


def _period_matches(unit: dict[str, Any], form: str) -> bool:
    duration = _duration_days(unit)
    if duration is None:
        return True
    return 250 <= duration <= 400 if form.startswith("10-K") else 65 <= duration <= 120


def select_financial_facts(payload: dict[str, Any], cik: str) -> list[FinancialFact]:
    us_gaap = payload.get("facts", {}).get("us-gaap", {})
    results: list[FinancialFact] = []
    seen: set[tuple[str, int, str]] = set()
    for registry_key, definition in CONCEPTS.items():
        for candidate in definition.candidates:
            concept = us_gaap.get(candidate)
            if not concept:
                continue
            for unit_name in definition.preferred_units:
                units = concept.get("units", {}).get(unit_name, [])
                for unit in sorted(
                    units,
                    key=lambda item: (item.get("filed", ""), item.get("accn", "")),
                    reverse=True,
                ):
                    form = unit.get("form", "")
                    fiscal_year = unit.get("fy")
                    accession = unit.get("accn")
                    if (
                        form not in {"10-K", "10-Q", "10-K/A", "10-Q/A"}
                        or not fiscal_year
                        or not accession
                    ):
                        continue
                    base_form = form.replace("/A", "")
                    key = (registry_key, int(fiscal_year), unit.get("fp", base_form))
                    if key in seen or not _period_matches(unit, base_form):
                        continue
                    source = (
                        filing_url(cik, accession, unit.get("frame", "filing") + ".htm")
                        if unit.get("frame")
                        else f"https://www.sec.gov/Archives/edgar/data/{int(cik)}/{accession.replace('-', '')}/"
                    )
                    results.append(
                        FinancialFact(
                            concept=registry_key,
                            displayLabel=definition.label,
                            value=float(unit["val"]),
                            unit=unit_name,
                            periodStart=date.fromisoformat(unit["start"])
                            if unit.get("start")
                            else None,
                            periodEnd=date.fromisoformat(unit["end"]),
                            fiscalYear=int(fiscal_year),
                            fiscalQuarter=int(unit["fp"][1:])
                            if str(unit.get("fp", "")).startswith("Q")
                            else None,
                            form=form,
                            filedDate=date.fromisoformat(unit["filed"]),
                            accession=accession,
                            sourceUrl=source,
                        )
                    )
                    seen.add(key)
            break
    return sorted(
        results, key=lambda fact: (fact.fiscalYear, fact.fiscalQuarter or 5, fact.concept)
    )


def calculate_growth(current: FinancialFact, previous: FinancialFact) -> float | None:
    if current.concept != previous.concept or previous.value == 0:
        return None
    return (current.value - previous.value) / abs(previous.value)


def calculate_margin(numerator: FinancialFact, revenue: FinancialFact) -> float | None:
    if (
        revenue.concept != "revenue"
        or revenue.value == 0
        or numerator.periodEnd != revenue.periodEnd
    ):
        return None
    return numerator.value / revenue.value
